### PR TITLE
Update to 3Dmol 2.0, now directly compatible with TypeScript

### PR DESCRIPTION
**Pull request by Luthaf**

### CODE REVIEW

**Comment by ceriottm** [comment]
You have asked to break this, and your wish has been granted. When I try to load the azaphenacene example, the npm server crashes and gives the error
```
<e> [webpack-dev-middleware] ModuleNotFoundError: Module not found: Error: Can't resolve '/home/michele/lavoro/code/chemiscope/node_modules/ts-loader/dist/stringify-loader.js' in '/home/michele/lavoro/code/chemiscope/src/structure'
<e>     at /home/michele/lavoro/code/chemiscope/node_modules/webpack/lib/Compilation.js:2017:28
<e>     at /home/michele/lavoro/code/chemiscope/node_modules/webpack/lib/NormalModuleFactory.js:817:13
<e>     at eval (eval at create (/home/michele/lavoro/code/chemiscope/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:10:1)
<e>     at /home/michele/lavoro/code/chemiscope/node_modules/webpack/lib/NormalModuleFactory.js:275:22
<e>     at eval (eval at create (/home/michele/lavoro/code/chemiscope/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:9:1)
<e>     at /home/michele/lavoro/code/chemiscope/node_modules/webpack/lib/NormalModuleFactory.js:448:22
<e>     at /home/michele/lavoro/code/chemiscope/node_modules/webpack/lib/NormalModuleFactory.js:122:11
<e>     at /home/michele/lavoro/code/chemiscope/node_modules/webpack/lib/NormalModuleFactory.js:654:23
<e>     at /home/michele/lavoro/code/chemiscope/node_modules/neo-async/async.js:2830:7
<e>     at done (/home/michele/lavoro/code/chemiscope/node_modules/neo-async/async.js:2925:13)
<e> resolve '/home/michele/lavoro/code/chemiscope/node_modules/ts-loader/dist/stringify-loader.js' in '/home/michele/lavoro/code/chemiscope/src/structure'
<e>   using description file: /home/michele/lavoro/code/chemiscope/package.json (relative path: ./src/structure)
<e>     using description file: /home/michele/lavoro/code/chemiscope/package.json (relative path: ./node_modules/ts-loader/dist/stringify-loader.js)
<e>       no extension
<e>         /home/michele/lavoro/code/chemiscope/node_modules/ts-loader/dist/stringify-loader.js doesn't exist
<e>       .js
<e>         /home/michele/lavoro/code/chemiscope/node_modules/ts-loader/dist/stringify-loader.js.js doesn't exist
<e>       as directory
<e>         /home/michele/lavoro/code/chemiscope/node_modules/ts-loader/dist/stringify-loader.js doesn't exist
```

File: src/structure/assignBonds.ts [hunk 1]
```diff
@@ -108,8 +109,10 @@ function areConnected(atom1: AtomSpec, atom2: AtomSpec) {
     else if (distSquared < 0.5) return false;
     // maybe duplicate position.
     else if (distSquared > maxsq) return false;
+    /**  FIXME: altLoc no longer exists on AtomSpec
```
Comment (rosecers) [inline src/structure/assignBonds.ts:112]: What's the significance here?

File: src/structure/grid.ts
Comment (rosecers) [inline src/structure/grid.ts:1]: Are these changes related to 3dmol?

File: src/structure/viewer.ts [hunk 2]
```diff
@@ -390,19 +396,16 @@ export class MoleculeViewer {
                 cstyle: { hidden: true },
             });
         }
-        assignBonds(this._current.model.selectedAtoms({}) as $3Dmol.AtomSpec[]);
-        this._current.model.addAtomSpecs(['index']);
+        assignBonds(this._current.model.selectedAtoms({}));
```
Comment (rosecers) [inline src/structure/viewer.ts:399]: Can we drop the explicit type casting in the new version?

**Review by rosecers — COMMENTED** [review]
So functionality-wise, I'm seeing little difference, but code-wise, I'm seeing a lot of changes that I can't quite follow -- can you give some insight into the API changes that necessitated these code changes?

File: src/structure/viewer.ts (hunk 2 above)
Comment (Luthaf) [reply src/structure/viewer.ts:399]: yes, the new type declarations are more accurate, so we don't have to do this manually anymore

File: src/structure/grid.ts
Comment (Luthaf) [reply src/structure/grid.ts:1]: No, just some cleanup on the go, renaming widget => viewer

File: src/structure/assignBonds.ts (hunk 1 above)
Comment (Luthaf) [reply src/structure/assignBonds.ts:112]: we don't actually use altLoc, this is a biomolecule thing. This function used to handle it since it was extracted out of 3dmol (it is not publicly accessible), I need to remove this bit of code

**Comment by Luthaf** [comment]
>  I'm seeing a lot of changes that I can't quite follow -- can you give some insight into the API changes that necessitated these code changes?

The API changes are minimal, mainly different types declaration and `index` now being an actual field for AtomSpec.

I also took the opportunuity to cleanup some of the code surrounding the grid and viewer classes.

File: src/structure/grid.ts
Comment (rosecers) [reply src/structure/grid.ts:1]: My favorite kind of inclusion in a pull request! /s

File: src/structure/assignBonds.ts (hunk 1 above)
Comment (rosecers) [reply src/structure/assignBonds.ts:112]: Okay! I might suggest leaving a comment that altloc is not supported, instead of a FIXME

**Comment by ceriottm** [comment]
Played a bit more with it, also on the jupyter side, and all seems in good order.

File: src/structure/assignBonds.ts (hunk 1 above)
Comment (Luthaf) [reply src/structure/assignBonds.ts]: I removed this comment, since this more confusing than helpful

File: src/structure/grid.ts
Comment (Luthaf) [reply src/structure/grid.ts:1]: well, nobody else is doing any code cleanup on this repo, so I'd rather include in random PR when I have time to work on chemiscope than not do it at all. This is the [boy scout approach to coding](https://stevenpcurtis.medium.com/the-boy-scout-rule-in-coding-a0b4d8130798), leaving the code cleaner than you found it.
